rosplan_sensing_interface/scripts/sensing_interface.py

```diff
@@ -0,0 +1,440 @@
+                        changed = True
+                    self.sensed_services[pred_name + ':' + params] = (result, changed)
+                    self.srv_mutex.release()
+
+            r.sleep()
+        return None  # Finish thread
+
+    def update_kb(self):
+        kus = KnowledgeUpdateServiceArrayRequest()
+        # Get info from KB functions and propositions (to know type of variable)
+        # Fill update
+        self.mutex.acquire(True)
+        sensed_predicates = self.sensed_topics.copy()
+        self.mutex.release()
+        self.srv_mutex.acquire(True)
+        sensed_predicates.update(self.sensed_services.copy())
+        self.srv_mutex.release()
+        for predicate, (val, changed) in sensed_predicates.iteritems():
+            if not changed:
+                continue
+
+            if predicate in self.sensed_topics:
+                self.mutex.acquire(True)
+                self.sensed_topics[predicate] = (self.sensed_topics[predicate][0], False)  # Set to not changed
+                self.mutex.release()
+            else:
+                self.srv_mutex.acquire(True)
+                self.sensed_services[predicate] = (self.sensed_services[predicate][0], False)  # Set to not changed
+                self.srv_mutex.release()
+            predicate_info = predicate.split(':')
+            ki = KnowledgeItem()
+            ki.attribute_name = predicate_info.pop(0)
+            if type(val) is bool:
+                ki.knowledge_type = ki.FACT
+                ki.is_negative = not val
+            else:
+                ki.knowledge_type = ki.FUNCTION
+                ki.function_value = val
+
+            # TODO what to do if no parameters specified? iterate over all the instantiated parameters and add it?
+            for i, param in enumerate(predicate_info):
+                kv = KeyValue()
+                kv.key = 'p' + str(i)
+                kv.value = param
+                ki.values.append(kv)
+
+            kus.update_type += np.array(kus.ADD_KNOWLEDGE).tostring()
+            kus.knowledge.append(ki)
+
+        # Update the KB with the full array
+        if len(kus.update_type) > 0:
+            try:
+                self.update_kb_srv.call(kus)
+            except Exception as e:
+                rospy.logerr("KCL (SensingInterface) Failed to update knowledge base: %s" % e.message)
+
+    def get_kb_attribute(self, attribute_name):
+        request = GetAttributeServiceRequest(attribute_name)
+        ret = self.get_state_propositions_srv.call(request)
+        if len(ret.attributes) > 0:
+            return ret.attributes
+        return self.get_state_functions_srv.call(request).attributes
+
+
+if __name__ == "__main__":
+    rospy.init_node('rosplan_sensing_interface', anonymous=False)
+    rps = RosplanSensing()
+    t = threading.Thread(target=RosplanSensing.call_services, args=(rps,))
+
+    # main loop
+    rate = rospy.Rate(10)  # 10hz
```

Comment:
Can this be parameterized in the launch file?

Reply:
Agree.